Compiled dag

stephanie-wang opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

## Why are these changes needed?

<!-- Please give a short summary of the change and the problem this solves. -->

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

Commit:
ip

Commit:
basic working.

Commit:
enhancement

Commit:
working now.

stephanie-wang committed:
Merge remote-tracking branch 'sang/dag-api' into compiled-dag

stephanie-wang committed:
Merge remote-tracking branch 'upstream/master' into compiled-dag

stephanie-wang committed:
tmp

Signed-off-by: Stephanie Wang <swang@cs.berkeley.edu>

stephanie-wang committed:
Merge branch 'mutable-objects-2' into compiled-dag

stephanie-wang committed:
scatter-gather DAG works

Signed-off-by: Stephanie Wang <swang@cs.berkeley.edu>

stephanie-wang committed:
fix

Signed-off-by: Stephanie Wang <swang@cs.berkeley.edu>

stephanie-wang committed:
fix

Signed-off-by: Stephanie Wang <swang@cs.berkeley.edu>

stephanie-wang committed:
compile?

Signed-off-by: Stephanie Wang <swang@cs.berkeley.edu>

stephanie-wang committed:
Merge branch 'mutable-objects-2' into compiled-dag

stephanie-wang committed:
unit test

Signed-off-by: Stephanie Wang <swang@cs.berkeley.edu>

stephanie-wang committed:
Merge branch 'mutable-objects-2' into compiled-dag

stephanie-wang committed:
tmp

Signed-off-by: Stephanie Wang <swang@cs.berkeley.edu>

ericl commented on python/ray/dag/dag_node.py:
```diff
@@ -103,9 +107,20 @@ async def get_object_refs_from_last_execute(self) -> Dict[str, Any]:
+        *args,
+        _ray_cache_refs: bool = False,
+        _ray_cache_actors: bool = True,
+        compiled: bool = False,
```
Shall we make an `experimental_compile()` method that returns a compiled node type instead of adding an arg here? I think the arg is a bit confusing since it's not clear that we cache the compiled DAG on the first interaction with execute.

stephanie-wang replied:
Yeah agree, I was thinking that too actually.

ericl commented on python/ray/dag/dag_node.py:
```diff
@@ -115,15 +130,31 @@ def execute(
 
+    def destroy_compiled_dag(self):
+        _, _, _, monitor = self.compiled()
+        monitor.destroy()
```
Drop the fault tolerance/cancellation stuff from this PR?

stephanie-wang replied:
Oops, missed this...

stephanie-wang committed:
Merge branch 'mutable-objects-2' into compiled-dag

stephanie-wang committed:
Revert "tmp"

This reverts commit 93968107d6de00336ca5af23a3c95e9f655ddc89.

stephanie-wang committed:
cleanup

Signed-off-by: Stephanie Wang <swang@cs.berkeley.edu>

stephanie-wang committed:
Support no-OutputNode DAGs

Signed-off-by: Stephanie Wang <swang@cs.berkeley.edu>

stephanie-wang committed:
Support non-DAG args

Signed-off-by: Stephanie Wang <swang@cs.berkeley.edu>

stephanie-wang committed:
errors

Signed-off-by: Stephanie Wang <swang@cs.berkeley.edu>

stephanie-wang committed:
lint

Signed-off-by: Stephanie Wang <swang@cs.berkeley.edu>

stephanie-wang committed:
doc

Signed-off-by: Stephanie Wang <swang@cs.berkeley.edu>

rkooo567 commented on python/ray/dag/compiled_dag_node.py:
```diff
@@ -0,0 +1,331 @@
+    return ray_channel.Channel(buffer_size_bytes, num_readers)
+
+
+def do_allocate_channel(
```
Can you comment somewhere the purpose of do_* methods? I think I get why you do this, but it is a little confusing

rkooo567 commented on python/ray/dag/compiled_dag_node.py:
```diff
@@ -0,0 +1,331 @@
+        # Add placeholders for input channels.
+        for inp in inputs:
+            if isinstance(inp, ray_channel.Channel):
+                input_channel_idxs.append((len(resolved_inputs), inp))
```
Nit, but maybe just 

```python3
for i, inp in enumerate(inputs):
    ...
    input_channel_idxs.append(i, inp)
```
?

rkooo567 commented on python/ray/dag/compiled_dag_node.py:
```diff
@@ -0,0 +1,331 @@
+                resolved_inputs.append(inp)
+
+        while True:
+            for idx, chan in input_channel_idxs:
```
```suggestion
            for idx, channel in input_channel_idxs:
```

rkooo567 commented on python/ray/dag/compiled_dag_node.py:
```diff
@@ -0,0 +1,331 @@
+            output_val = method(*resolved_inputs)
+
+            self._output_channel.write(output_val)
+            for _, chan in input_channel_idxs:
```
```suggestion
            for _, channel in input_channel_idxs:
```

rkooo567 commented on python/ray/dag/tests/test_accelerated_dag.py:
```diff
@@ -0,0 +1,134 @@
+        for chan in output_channels:
+            chan.end_read()
+
+
```
can we add a test with more than 1 channel passed to a single dag node?

rkooo567 commented on python/ray/dag/compiled_dag_node.py:
```diff
@@ -0,0 +1,331 @@
+                chan.end_read()
+
+    except Exception as e:
+        print("Task aborted", e)
```
logger.exception instead?

rkooo567 replied:
actually do we really need this try except? Seems not very useful..

stephanie-wang replied:
Mainly just to make sure that we log the exception. Since the task is infinite, the driver may be stuck in some ray.get waiting for a channel output.

rkooo567 commented on python/ray/dag/compiled_dag_node.py:
```diff
@@ -0,0 +1,331 @@
+        # Cached attributes that are set during compilation.
+        self.dag_input_channel = None
+        self.dag_output_channels = None
+        self.node_idx_to_output_channels = {}
```
is it needed? Isn't it just idx_to_task[idx].output_channel?

rkooo567 commented on python/ray/dag/compiled_dag_node.py:
```diff
@@ -0,0 +1,331 @@
+        self.dag_input_channel = None
+        self.dag_output_channels = None
+        self.node_idx_to_output_channels = {}
+        self.worker_task_refs = []
```
can you add comment what this means?

rkooo567 commented on python/ray/dag/compiled_dag_node.py:
```diff
@@ -0,0 +1,331 @@
+        self.node_idx_to_output_channels = {}
+        self.worker_task_refs = []
+
+    def _add_node(self, node):
```
Add type annotation!

rkooo567 commented on python/ray/dag/compiled_dag_node.py:
```diff
@@ -0,0 +1,331 @@
+    """Wraps the normal Ray DAGNode with some metadata."""
+
+    def __init__(self, idx, dag_node: "ray.dag.DAGNode"):
+        self.idx = idx
```
can you add types for variables here and add comments for attributes?

rkooo567 replied:
e.g., it is confusing if self.args is for CompiledTask vs DagNode

rkooo567 commented on python/ray/dag/dag_node.py:
```diff
@@ -103,9 +107,21 @@ async def get_object_refs_from_last_execute(self) -> Dict[str, Any]:
+        is cached.
+        """
+        if self._compiled_dag is None:
+            self._compiled_dag = build_compiled_dag_from_ray_dag(self)
```
Isn't this creating circular dependency? (self -> _compiled_dag and _compiled_dag has reference to self). Maybe we should use weakreference

stephanie-wang replied:
I think this is fine since the DAG itself should use very little memory.

stephanie-wang replied:
We also don't want to have cases where the original dag gets dropped and now the compiled DAG no longer works.

rkooo567 commented on python/ray/dag/dag_node.py:
```diff
@@ -218,7 +234,7 @@ def _apply_and_replace_all_child_nodes(
         )
 
-    def apply_recursive(self, fn: "Callable[[DAGNode], T]") -> T:
+    def apply_recursive(self, fn: "Callable[[DAGNode], T]", cache=None) -> T:
```
cache -> it is not used?

rkooo567 commented on python/ray/dag/compiled_dag_node.py:
```diff
@@ -0,0 +1,331 @@
+        """Compile an execution path. This allocates channels for adjacent
+        tasks to send/receive values. An infinite task is submitted to each
+        actor in the DAG that repeatedly receives from input channel(s) and
+        sends to output channel(s).
```
Can you mention the API is idempotent?

rkooo567 commented on python/ray/dag/compiled_dag_node.py:
```diff
@@ -0,0 +1,331 @@
+        if self.input_task_idx is None:
+            self._preprocess()
+
+        if self.dag_input_channel is not None and self.dag_output_channels is not None:
```
is there a case one of them is None, and the other is not? If there's no such case, can you add assert?

rkooo567 commented on python/ray/dag/dag_node.py:
```diff
@@ -103,9 +107,21 @@ async def get_object_refs_from_last_execute(self) -> Dict[str, Any]:
     def clear_cache(self):
         self.cache_from_last_execute = {}
 
+    def experimental_compile(self) -> "ray.dag.CompiledDAG":
```
it's NIT, but why don't we do 

ray.compile(dag)

instead of 

dag.compile()?

This way ,we can avoid circular dependencies (and also circular import that we have to workaround it by importing within a method)

ericl replied:
Let's do this later I think? I don't think the dependency issue we need to worry about.

rkooo567 commented on python/ray/dag/compiled_dag_node.py:
```diff
@@ -0,0 +1,331 @@
+        # For each task node, set its upstream and downstream task nodes.
+        for idx, task in self.idx_to_task.items():
+            dag_node = task.dag_node
+            if not (
```
nit,

```python3
if (isinstance(dag_node, InputNode)
    or isinstance(dag_node, OutputNode)
    or isinstance(dag_node, ClassMethodNode):
    continue
```

rkooo567 replied:
(so that we can avoid nesting)

rkooo567 commented on python/ray/dag/compiled_dag_node.py:
```diff
@@ -0,0 +1,331 @@
+        self.idx = idx
+        self.dag_node = dag_node
+
+        self.args = []
```
Is this actually necessary? Can we just create a property

```
@property
def args(self):
    return self.dag_node.get_args()
```
?

rkooo567 commented on python/ray/dag/compiled_dag_node.py:
```diff
@@ -0,0 +1,331 @@
+
+        # Find the (multi-)output node to the DAG.
+        for idx, task in self.idx_to_task.items():
+            if len(task.dependent_node_idxs) == 0:
```
NIT: this could be included in the prvious loop?

```python3
for idx, task in self.idx_to_task.items():
            if isinstance(task.dag_node, InputNode):
                assert self.input_task_idx is None, "more than one InputNode found"
                self.input_task_idx = idx
            if len(task.dependent_node_idxs) == 0:
                assert self.output_task_idx is None, "More than one output node found"
                self.output_task_idx = idx
```

stephanie-wang replied:
I just did this for readability. The performance here doesn't matter much since we only do it once.

rkooo567 commented on python/ray/dag/compiled_dag_node.py:
```diff
@@ -0,0 +1,331 @@
+
+        # Find the (multi-)output node to the DAG.
+        for idx, task in self.idx_to_task.items():
+            if len(task.dependent_node_idxs) == 0:
```
This part seems a bit confusing to me. 

Isn't dependent_node == arguments? Why does output node have 0 dependent node id?

stephanie-wang replied:
Ah it's a reverse index for arguments. dependent_node is downstream tasks. I'll call it "downstream_node_idxs" name to make it clearer.

rkooo567 commented on python/ray/dag/compiled_dag_node.py:
```diff
@@ -0,0 +1,331 @@
+                )
+
+            # Assign the task with the correct input and output buffers.
+            worker_fn = task.dag_node._get_remote_method("__ray_call__")
```
Maybe we can make a wrapper? 

```python3
def invoke_remote_method_from_task(task, func, *args, **kwargs):
    worker_fn = task.dag_node._get_remote_method("__ray_call__")
    return worker_fn.remote(func, args, kwargs)
```
something like this?

stephanie-wang replied:
I think it's fine not to do.

rkooo567 commented on python/ray/dag/compiled_dag_node.py:
```diff
@@ -0,0 +1,331 @@
+"""
+
+
+class CompiledDAG:
```
Should we add tests that validate DAG itself (without running execute)? 

1. Convert regular DAG and make sure returned compiled dag contains all regular DAG nodes + correct worker_task_refs
2. Test cases where exceptions are raised

stephanie-wang replied:
Makes sense, I'll do 1.

2 is already done (see test_dag_errors).

rkooo567 reviewed (commented):
Q: is the perf benchmark result similar or the same from before?

stephanie-wang committed:
Merge branch 'mutable-objects-2' into compiled-dag

angelinalg committed:
[docs] [data] remove Data Processing tag from sidenav and examples (#41686)

After much discussion, the Data Processing tag is too general. Removing it from the Example Gallery.
cc: @amogkam @scottjlee @pcmoritz

Signed-off-by: angelinalg <122562471+angelinalg@users.noreply.github.com>

sven1977 committed:
[RLlib] Add and enhance fault-tolerance tests for APPO. (#40743)

angelinalg committed:
[docs][serve] Fix subsection title hierarchy (#41746)

Subsections are appearing in the Serve Examples page as separate docs vs sections of the Streaming doc.

Signed-off-by: angelinalg <122562471+angelinalg@users.noreply.github.com>

GeneDer committed:
[Serve] allow gRPC deployment to use grpc context (#41667)

This PR passes a grpc_context to deployments if the deployment uses it to get gRPC request related info can use to set code, details, trailing metadata, and compression. The original grpc._cython.cygrpc._ServicerContext type is not serializable, so we created a RayServegRPCContext to be able to pass to the deployment. Will follow up with doc change.

---------

Signed-off-by: Gene Su <e870252314@gmail.com>

edoakes committed:
[serve] Update replica ID format (and other random strings) to better match UUIDs (#41737)

Small cosmetic update to our replica ID generation to make it use the same alphabet as UUID (which is standard and we use for request IDs).

---------

Signed-off-by: Edward Oakes <ed.nmi.oakes@gmail.com>

sihanwang41 committed:
[Serve] Update system_logging_config to logging_config (#41749)

Signed-off-by: Sihan Wang <sihanwang41@gmail.com>

alexeykudinkin committed:
[Serve] Patching ActorProxyWrapper to properly handle `is_drained` RPC (#41744)

This is a minified version of https://github.com/ray-project/ray/pull/41722/files, specifically put to be cherry-picked into 2.9

Addresses #41726

---------

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

rynewang committed:
[core] retryable exceptions for method (#41194)

Implements retryable exceptions for methods. Also plumbs direct_actor_task_submitter quite a bit. Behavior:

Added new method-level annotation max_retries that overrides actor's max_task_retries if any.
If user_exceptions=True | List[exception class types] and ((max_retries or max_task_retries) > 0) we may retry the method by issuing another task invocation to the actor.
Both exception-retry and actor-death-retry counts toward max_retries. For example for a max_retries=3 call, and there are 2 actor deaths and 1 exception and 1 return, we can return the value.
For a streaming generator call, if it yielded 4 values then raises an exception, we retry by calling the method again and ignoring the first 4 values and start yielding the 5th value in the second run.
Java and CPP: they still have max_task_retries on actor deaths, but I did not add max_retries or retry_exceptions.

jjyao committed:
[Core] Fix core worker client pool leak (#41535)

Currently core worker client pool doesn't remove clients in most cases (there are one or two places where Disconnect() might be called) and this caused memory leak. This PR adds a GC inside core worker client pool to remove IDLE clients (i.e.g clients that don't have active connections).

Signed-off-by: Jiajun Yao <jeromeyjj@gmail.com>

rkooo567 committed:
[Core] Fix check failure when sync and async tasks are mixed up (#41724)

When sync and async tasks are mixed up, there are sometimes where TaskID is not correctly set, which causes the check failure. This is the case where check failure correctly found a bug.

I fixed the issue by passing TaskID correctly. Also improved get_current_task_id API to always return the correct task ID when used within raylet.pyx

jjyao committed:
[Core] Speed up single_client_tasks_sync (#41743)


---------

Signed-off-by: Jiajun Yao <jeromeyjj@gmail.com>

rickyyx committed:
[core][task-backend] Add unit testing for gcs task manager on job finishes. (#41687)



---------

Signed-off-by: rickyyx <rickyx@anyscale.com>

can-anyscale committed:
[ci][wins/3] add --operating-system flag to rayci (#41636)

Add --operating-system flag. Either use LinuxTesterContainer or WindowsTesterContainer based on the flag. The only if condition!

Signed-off-by: can <can@anyscale.com>

stephanie-wang committed:
cleanup

Signed-off-by: Stephanie Wang <swang@cs.berkeley.edu>

stephanie-wang committed:
cleanup

Signed-off-by: Stephanie Wang <swang@cs.berkeley.edu>

stephanie-wang committed:
perf

Signed-off-by: Stephanie Wang <swang@cs.berkeley.edu>

peytondmurray committed:
[Doc] Fix the linkcheck CI job (#41598)

- This PR fixes the linkcheck job by bypassing ci.sh, and using a docker container on the buildkite runner.
- This PR also fixes a ton of broken links that were causing linkcheck to fail (even without rate limit issues), mostly by removing links to examples that were deleted.
- I also removed the global depends_on in lint.rayci.yml; not all lint jobs require the same environment.

Signed-off-by: pdmurray <peynmurray@gmail.com>
Signed-off-by: Peyton Murray <peynmurray@gmail.com>
Co-authored-by: angelinalg <122562471+angelinalg@users.noreply.github.com>

can-anyscale committed:
[ci] turn off verbose mode when zipping logs (#41760)

As title, the tar command is spamming a lot of logs

Signed-off-by: can <can@anyscale.com>

can-anyscale committed:
[ci] always update apt in dashboard tests (#41754)

As title, the test might fail if apt has been cached and not updated. Currently blocking release branch https://buildkite.com/ray-project/postmerge/builds/2005#018c4a89-ec04-4f3b-83de-744b6e85fe63/159-895

Signed-off-by: can <can@anyscale.com>

jiwq committed:
[CI] Clean the dev env after build wheels for macOS (#41694)

deactivate the related conda env.

Signed-off-by: Wanqiang Ji <me@jiwq.cn>
Signed-off-by: Lonnie Liu <95255098+aslonnie@users.noreply.github.com>
Co-authored-by: Lonnie Liu <95255098+aslonnie@users.noreply.github.com>

sven1977 committed:
[RLlib] Fix `policy_to_train` logic for new API stack generically for all algorithms. (#41529)

rkooo567 committed:
[Core][CI] Mark test_runtime_env_working_dir_3 flaky

Co-authored-by: SangBin Cho <sangcho@sangcho-LT93GQWG9C.local>

stephanie-wang committed:
[core] Support mutable plasma objects (#41515)

This is a first pass on introducing an experimental "channel" concept that can be used for direct worker-worker communication, bypassing the usual Ray Core components like the driver and raylet.

Channels are implemented as mutable plasma objects. The object can be written multiple times by a client. The writer must specify the number of reads that can be made before the written object value is no longer valid. Reads block until the specified version or a later one is available. Writes block until all readers are available. Synchronization between a single writer and multiple readers is performed through a new header for plasma objects that is stored in shared memory.

API:

    channel: Channel = ray.experimental.channel.Channel(buf_size): Client uses the normal ray.put path to create a mutable plasma object. Once created and sealed for the first time, the plasma store synchronously reads and releases the object. At this point, the object may be written by the original client and read by others.
    channel.write(val): Use the handle returned by the above to send a value through the channel. The caller waits until all readers of the previous version have released the object, then writes a new version.
    val = channel.begin_read(): Blocks until a value is available. Equivalent to ray.get. This is the beginning of the client's read.
    channel.end_read(): End the client's read, marking the channel as available to write again.

---------

Signed-off-by: Stephanie Wang <swang@cs.berkeley.edu>

stephanie-wang committed:
add normal DAG

Signed-off-by: Stephanie Wang <swang@cs.berkeley.edu>

stephanie-wang committed:
Merge remote-tracking branch 'upstream/master' into compiled-dag

stephanie-wang committed:
x

Signed-off-by: Stephanie Wang <swang@cs.berkeley.edu>

stephanie-wang commented:
Closing in favor of #41769.

ericl committed:
[core] Add an internal system concurrency group for executing compiled dag tasks

As part of https://github.com/ray-project/enhancements/pull/48/files, we need the ability to launch system-managed tasks to execute compiled DAG work. These need to run in a separate concurrency group so they do not block other actor tasks.

This is an internal/experimental API only.

aslonnie committed:
[lint] format build files (#41766)

Signed-off-by: Lonnie Liu <lonnie@anyscale.com>

vitsai committed:
[core] Small redis context cleanup (#41776)

Cleanup of dead code in redis context.

Signed-off-by: vitsai <vitsai@cs.stanford.edu>

aslonnie committed:
[lint] build file format for src dir (#41767)

Signed-off-by: Lonnie Liu <lonnie@anyscale.com>

jjyao committed:
Revert "[core] Support mutable plasma objects (#41515)" (#41784)

This reverts commit cb5bb4e7763994db4f4e765ce9dc74376d7eedca.

can-anyscale committed:
[ci] unbreak master (#41793)

- Move linkcheck to skip-on-premerge, currently failing on master
- Move [CI] linux://rllib:learning_tests_cartpole_crashing_and_stalling_appo is failing/flaky on master. #41792 to flaky; almost always failing now

Signed-off-by: can <can@anyscale.com>

stephanie-wang committed:
Merge branch 'mutable-objects-2' into compiled-dag

bveeramani committed:
[Data] Fix unused `columns` parameter in `read_parquet_bulk` (#41806)

#40900 refactored the read_parquet_bulk implementation and introduced a regression where the columns parameter wasn't used. This PR fixes the bug.

Signed-off-by: Balaji Veeramani <balaji@anyscale.com>

vitsai committed:
[core] Refactor Redis code (#41771)

- Make redis_context_ unique_ptr instead of raw
- Clean up some templating so smart pointers can have deleters attached instead of leak-prone manual deletion
- Switch to return structures instead of return arguments

Signed-off-by: vitsai <vitsai@cs.stanford.edu>

raulchen committed:
[data] Optimize OpState.outqueue_num_blocks (#41748)

Signed-off-by: Hao Chen <chenh1024@gmail.com>

edoakes committed:
[serve] Remove `controller_name` across all components (#41733)

Removing dead code; this was previously required to support detached=False mode.

---------

Signed-off-by: Edward Oakes <ed.nmi.oakes@gmail.com>

fishbone committed:
Remove unnecessary test file in wheel (#41775)

This PR remove the test file in pip wheel which lead to the ray start failed

can-anyscale committed:
[ci] upload build data to go/flaky (#41796)

For a few build that doesn't use rayci, need to add code to upload build data to go/flaky

Signed-off-by: can <can@anyscale.com>

bveeramani committed:
[Data] Add example of how to read and write custom file types (#41785)

#40127 removed the "Implementing a Custom Datasource" example because it used deprecated APIs. This PR introduces a new example that uses up-to-date APIs.

---------

Signed-off-by: Balaji Veeramani <balaji@anyscale.com>

rkooo567 committed:
[Core] Improve DAG API to support tensor parallel DAG (#41231)

Support OutputNode
Allow to create bind from regular actor. It is needed because
Actor needs to be reused
Currently, you can have only 1 DAG per actor because the actor is a starting point of the DAG. This allows us to
make a task as a starting node of the DAG instead of the actor
This also allows to have more than one InputNode per each actor
This PR also removes the unused code

stephanie-wang committed:
Re-merge mutable objects (#41515) (#41789)

See #41515.

This updates to only compile new code on linux. OSX does not support shared memory semaphores, only named semaphores.

---------

Signed-off-by: Stephanie Wang <swang@cs.berkeley.edu>

stephanie-wang committed:
Merge remote-tracking branch 'upstream/master' into compiled-dag

stephanie-wang committed:
merge

Signed-off-by: Stephanie Wang <swang@cs.berkeley.edu>

stephanie-wang committed:
revert

Signed-off-by: Stephanie Wang <swang@cs.berkeley.edu>

stephanie-wang committed:
revert

Signed-off-by: Stephanie Wang <swang@cs.berkeley.edu>

stephanie-wang committed:
x

Signed-off-by: Stephanie Wang <swang@cs.berkeley.edu>

stephanie-wang committed:
buffer size bytes

Signed-off-by: Stephanie Wang <swang@cs.berkeley.edu>

angelinalg committed:
[docs][serve][tune] Remove serve/tutorial/rllib and tune-sklearn examples (#41758)

Purging outdated or low-value examples from the Example Gallery.
@pcmoritz @richardliaw

---------

Signed-off-by: angelinalg <122562471+angelinalg@users.noreply.github.com>

aslonnie committed:
[lint] buildifier format train build files (#41833)

Signed-off-by: Lonnie Liu <lonnie@anyscale.com>

aslonnie committed:
[lint] buildifier format tune docs build files (#41834)

formatting only

Signed-off-by: Lonnie Liu <lonnie@anyscale.com>

angelinalg committed:
remove duplicate of huggingface_vit_batch_prediction (#41738)

Part of a larger effort to better curate examples. This example appears twice, once as a Code Example, and a second time as a Tutorial. I thought that was redundant so I removed the duplication.

Signed-off-by: angelinalg <122562471+angelinalg@users.noreply.github.com>

zcin committed:
[serve] Fix bug in deployment state machine (#41799)

If a deployment is autoscaling and replicas take a long time to start, there is a bug that makes the state transition to (UPDATING, AUTOSCALING) which is a combination that should never occur. Instead, we should just update the message but not the status.

---------

Signed-off-by: Cindy Zhang <cindyzyx9@gmail.com>

justinvyu committed:
[doc][train] Clarify error message when trying to use local storage for multi-node distributed training and checkpointing (#41832)

Ray 2.7 removed support for using the head node as the persistent storage for checkpoints and artifacts in a multi-node distributed training. The alternative recommendation is to use cloud storage or a shared filesystem instead via `RunConfig(storage_path)`.

Ray Train/Tune will error if the user attempts to checkpoint `ray.train.report(..., checkpoint=...)` from a worker that's on a remote node. This is because the new assumption is that all worker nodes have access to read/write from the same persistent storage, and the "head node local storage" is not accessible by all nodes.

However, the error message that shows up is confusing. All nodes can technically access the local path in the message -- the problem is that not all nodes can access the the SAME local path. This PR improves the error message to make this more clear and to suggest an actionable fix. This PR also updates most of the getting started user guides to mention the multi-node storage requirement and links to the storage user guide.

---------

Signed-off-by: Justin Yu <justinvyu@anyscale.com>

justinvyu committed:
[doc][train] Clarify `prepare_data_loader` shuffle behavior and include `set_epoch` usage in all examples (#41807)

`prepare_data_loader` adds a `DistributedSampler` to an existing pytorch `DataLoader` object. To do this, it recreates a `DataLoader` object and passes most arguments through from the original object, but also makes some implicit assumptions that are not configurable/visible to the caller.

For example, if using just vanilla pytorch by itself, it's possible to do: `train_dataloader = DataLoader(..., shuffle=False, sampler=DistributedSampler(shuffle=True))`. Here, the `DataLoader` sets `shuffle=False`, but the `DistributedSampler` will still do a shuffle on every epoch so that the training data order is not always the same. The `shuffle=False` argument of the `DataLoader` is pretty much ignored because a custom sampler is supplied.

**However, with Ray Train, since this `prepare_data_loader` utility injects the `DistributedSampler` for the user, there's no visibility on the `shuffle` parameter.** Ray Train will detect the `shuffle` parameter set on the *original* dataloader, then pass that along to the `DistributedSampler`. So, it's not possible to have this `False+True` situation.

**Additionally, if `shuffle=True`, `DistributedSampler.set_epoch` must be called at the start of each epoch in order for the dataset ordering to be different for all workers *on every epoch.*** This is because the seed of the sampler is determined at the epoch start (`epoch seed = base random seed + epoch number`).

Shuffling can be very important for training a model successfully -- if the data order remains the same every epoch, it's possible that training never converges (ex: we ran into this issue training resnet18 on imagenet).

Signed-off-by: Justin Yu <justinvyu@anyscale.com>

GeneDer committed:
[Serve] Add gRPC context related docs (#41783)

Adding example code and descriptions for using gRPC context in Serve.

---------

Signed-off-by: Gene Su <e870252314@gmail.com>
Signed-off-by: Gene Der Su <gdsu@ucdavis.edu>
Co-authored-by: Edward Oakes <ed.nmi.oakes@gmail.com>

stephanie-wang committed:
optional

Signed-off-by: Stephanie Wang <swang@cs.berkeley.edu>

stephanie-wang committed:
x

Signed-off-by: Stephanie Wang <swang@cs.berkeley.edu>

stephanie-wang committed:
Merge remote-tracking branch 'upstream/master' into compiled-dag

stephanie-wang committed:
fix (#41854)

Un-break CI.

Signed-off-by: Stephanie Wang <swang@cs.berkeley.edu>

stephanie-wang committed:
Merge remote-tracking branch 'upstream/master' into compiled-dag

stephanie-wang committed:
x

Signed-off-by: Stephanie Wang <swang@cs.berkeley.edu>

stephanie-wang committed:
lint?

Signed-off-by: Stephanie Wang <swang@cs.berkeley.edu>

stephanie-wang committed:
test

Signed-off-by: Stephanie Wang <swang@cs.berkeley.edu>

stephanie-wang committed:
API

Signed-off-by: Stephanie Wang <swang@cs.berkeley.edu>

stephanie-wang committed:
x

Signed-off-by: Stephanie Wang <swang@cs.berkeley.edu>
